### DISCUSSION TITLE
Understanding startTransition() with NextJS Router

### DISCUSSION THREAD

**Discussion by osdiab** [vercel/next.js#54157] [Help] [answered]
### Summary

This is kind of random and more of a general React question, but I don't understand how code like the snippet here (https://github.com/vercel/next.js/blob/fc3bf373c5aba9c4e21c7365276ee97c703ee9a4/test/e2e/app-dir/navigation/app/external-push/%5BstorageKey%5D/page.js#L61-L63) works:

```
  let [isPending, startTransition] = useTransition()

// ...
          startTransition(() => {
            router.push('https://example.vercel.sh/stuff?abc=123')
          })
```

Specifically, how does React know when the page navigation has completed and therefore set `isPending` to false? Does it wait for Suspense boundaries to all resolve and just somehow knows that code totally disconnected from the actual invocation from the `startTransition()` invocation are related? I don't really follow how this works mechanically so it feels weird to just assume that it works exactly how I'd expect it to. Thanks!

**Comment by icyJoseph** [comment]
I think people often hang too much on the wrong side of the purpose of a transition.

You use these to mark an state update as non-blocking. Meaning that, if there's other state changes going on, React favours working on those, and while it does that, the pending flag is set. 

- https://react.dev/reference/react/useTransition

> A function that updates some state by calling one or more [set functions.](https://react.dev/reference/react/useState#setstate) React immediately calls scope with no parameters and marks all state updates scheduled synchronously during the scope function call as transitions. They will be [non-blocking](https://react.dev/reference/react/useTransition#marking-a-state-update-as-a-non-blocking-transition) and [will not display unwanted loading indicators.](https://react.dev/reference/react/useTransition#preventing-unwanted-loading-indicators)

Now, for a react state based router that's pretty good because, as said here https://react.dev/reference/react/useTransition#building-a-suspense-enabled-router, once you batch the router navigation/change, you can do something else, without waiting for the routing work to be complete.

The App Router is fully integrated into React's state management, that's why we can use transitions.

There is also the benefit of transitions being native to React, so that if there's a suspenseful update, in an already rendered component, you can instead use a transition to run the update, and you can use the `isPending` flag to show visual feedback while the suspenful update finishes: 

- https://react.dev/reference/react/useTransition#preventing-unwanted-loading-indicators
- https://react.dev/reference/react/Suspense#preventing-already-revealed-content-from-hiding

**Reply by osdiab** [reply]
Thanks for the message - I think I’m getting closer to understanding this but am not quite there.

> if there's other state transitions going on, React favours working on those, and while it does that, the pending flag is set.

I follow this, but then in the example of a page transition with some UI when isPending is true,

- the transition kicks off the router push. This is deferred
- isPending is set to true, this is not deferred
- showing the loading indicator as well is not deferred
- At this point all the non-deferred work is done. Taking your description at face value I expect isPending would be flipped back to false, but that doesn’t seem to be the case - isPending stays true until the route changes and the server component is fully loaded for some reason.
- So this gets to the idea that

> The App Router is fully integrated into React's state management, that's why we can use transitions.

- so is the idea that router.push does a lot of different things asynchronously including changing the navbar URL, triggering the server component fetch, etc
- and that somehow React knows that not only the literal call to router.push but also any subsequent updates that it would trigger for its lifetime are also part of this transition and therefore isPending should be true
- And it also understands when its lifetime is complete somehow?

this is what I fail to understand. In another circumstance,

- if I do some long running work in a transition that actually doesn’t result in a change to state nor what gets rendered at all, would it just be pending forever? Or does React listen to the lifetime of the callback or something to know when it’s “done”?

**Reply by icyJoseph** [reply]
> - At this point all the non-deferred work is done. Taking your description at face value I expect isPending would be flipped back to false, but that doesn’t seem to be the case - isPending stays true until the route changes and the server component is fully loaded for some reason.

Not following here. If you router push the rendered component tree changes right? Unless your component is global to at least both origin and destination route. I guess pushing to the same exact route, including queries, should bail out? And then you'd be able to see the flag set to false?


> - And it also understands when its lifetime is complete somehow?

The way I read it, the function passed to startTransition is immediately called, and state updates resulting from that are, let's say, stored at the React's work queue in a special way, processing them, while favoring other work that comes, that was not wrapped by startTransition. While it does that, isPending is true. Once it completes the work associated with the transition, it can yield the isPending flag back to false.

> - if I do some long running work in a transition that actually doesn’t result in a change to state nor what gets rendered at all, would it just be pending forever? Or does React listen to the lifetime of the callback or something to know when it’s “done”?

If no state changes are batched, then there's nothing pending, but there'll be a frame where the flag is true and then it'll be immediately set to false. A bit counter intuitive. 

Also you should only wrap the state updates on transitions:

- https://react.dev/reference/react/useTransition#react-doesnt-treat-my-state-update-as-a-transition

For example the timer has to be inverted so that the transition is called when the timer fires. Also look at the Promise example. 

The idea here is to collect the result of doing some work, and then execute the UI updates without blocking interactivity. When you set state, various components re-render and previously React would block until that was totally reflected in the UI. 

This is specially true of routing. There's likely, depending on the app, a lot of React work associated with changing a route, which should not block interactivity of unrelated components.

**Reply by osdiab** [reply]
What I’m saying is that router.push is not synchronous right? The typescript types even return a promise. It does a lot of things, including an async server fetch, and yet somehow when wrapping it in startTransition, isPending stays true until the full navigation including the server component render concludes as far as I understand. If startTransition contents need to be synchronous for React to pick them up then I don’t get why startTransition would work for route changes, given that it’s not synchronous as far as I can tell.

The idea that other work gets prioritized is fine to me, what I don’t get is how the transition knows to not conclude until all the async stuff a route change entails is complete.

**Reply by icyJoseph** [reply, marked as answer]
> What I’m saying is that router.push is not synchronous right? The typescript types even return a promise. 

Mmm no? The old router did that. The one you import from `next/router`, which only works for Pages Router. And yeah there you could await the router.push/replace calls.

The App Router is a whole re-work, and calling push/replace returns `void`, not a promise. It is the one you import from `next/navigation`.

What's a bit funky is that those methods are in turn wrapping a `navigate` call with a transition. So in the original example you shared, there's a transition within a transition.

The `navigate` function in turn uses a `useReducer` dispatch. As said, it is deeply integrated with React. The old Pages Router did its own thing. In part this is why the Router.events have disappeared. 

You can even, for now, use the ReduxDevTools to see what the router is doing. 
 
![Screenshot 2023-08-18 at 18 18 51](https://github.com/vercel/next.js/assets/21013447/84cb7e8f-fa4b-44e4-8f5e-13c379a8ce25)

**Reply by osdiab** [reply]
Makes sense, that’s helpful to follow!

**Reply by hongz1** [reply]
I use 'startTransition' callback when form submit button requires loading until router process finished.
As you see codes below, LoadingButton will stop spinning when isSubmitting flag is switched to false.
This behavior might cause clicking the button while redirecting page.
Therefore, adding 'isPending' flag from transition will ensure to keep loading status on the button.

This is what I realized when I wrote form component with loading button.

```
const router = useRouter();
const [isPending, startTransition] = useTransition();
import LoadingButton from '@mui/lab/LoadingButton';

const onSubmit = handleSubmit(async (data) => {
   try {
      const success = await userSignIn(data);
      if (success) {
         startTransition(() => {
            router.replace(paths.protectedRoot, { scroll: false });
         });
      } else {
         setError('email', 'Invalid Login Credentials.' });
      }
   } catch (err) {
      setError('email', { message: err?.message || 'Failed to Verify' });
   }
}

return (
   <>
      ...
      <FormProvider methods={methods} onSubmit={onSubmit}>
         <Stack spacing={3}>
            <LoadingButton fullWidth type="submit" color="inherit" loading={isSubmitting || isPending}>LogIn</LoadingButton>
         </Stack>
      </FormProvider>
   </>
)
```
